### PR TITLE
Fix orientation constraints failing on satisfied configurations

### PR DESCRIPTION
**Pull request by JafarAbdi**
### Description

After applying the following fix https://github.com/ros-planning/moveit/pull/2414 a pipeline I’m working on isn’t working anymore (with a bunch of errors about constraints not statisfied one of them is `ros.moveit_planners_ompl.constrained_goal_sampler: More than 80% of the sampled goal states fail to satisfy the constraints imposed on the goal sampler. Is the constrained sampler working correctly?`)

The reason is, right now we're calculating the error rotation matrix and call eulerAngles from Eigen to find the angles’ errors which return them in the following ranges `[0:pi]x[-pi:pi]x[-pi:pi]` ([link](https://eigen.tuxfamily.org/dox/group__Geometry__Module.html#ga17994d2e81b723295f5bc3b1f862ed3b))

The issues are
1- As shown in the first test, despite the error is 1.57rad around Z-axis, eulerAngles will return x=3.139981, y=3.139841, z=1.570810 causing the constraint to fail
2- Another example in the second test `0.15, -pi/2, 0.15` eulerAngles will return `3.05343, -1.5708, 3.05343` which again will cause the constraint to fail ([link](https://gcc.godbolt.org/z/4fvd5z))
3- This function isn’t stable around the identity rotation [link#1](https://github.com/stack-of-tasks/pinocchio/pull/1158#issuecomment-615105106) and [link#2](https://gcc.godbolt.org/z/Ycboa8)

~Even if we use the function from `unsupported/Eigen` it will cause similar problems but in different orientations for example if the error rotation matrix correspond to pure rotation about Y with angles bigger than pi/2 the returned roll/yaw angles will cause a satisfied constraint to fails again~

~To fix this I used the Euler angles of the desired and the current orientations to find the angles’ errors~

To fix this I did copy the implementation from [unsupported/Eigen/EulerSystem.h](https://gitlab.com/libeigen/eigen/-/blob/master/unsupported/Eigen/src/EulerAngles/EulerSystem.h#L187) since it’s not released yet which will return the angles in the following ranges `[-pi:pi]x[-pi/2:pi/2]x[-pi:pi]` and handled the singularities differently when we compare the error with the tolerances

~TODO: I need to add more tests before merging this~

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Comment by mamoll** [comment]
@JeroenDM this is relevant for you, too.

**Review by mamoll — APPROVED** [review]
This is some nice detective work, @JafarAbdi!

**Comment by JeroenDM** [comment]
@JafarAbdi Wow, thank you for figuring this out!

Since #2414 was merged my tests for #2402 failed. I suspect you also found the solution to my problem :)
I used some hardcoded test data [here](https://github.com/JeroenDM/moveit/blob/49df9efcbf81cff25ab6875023d57931869ed773/moveit_core/kinematic_constraints/test/test_orientation_constraints.cpp#L52), generated with a [Python script](https://gist.github.com/JeroenDM/f8ba5f1a60821c4cacf7c137eff6f409). If you think this is a good approach I can add the test for Euler angles to this PR.

File: moveit_core/kinematic_constraints/src/kinematic_constraint.cpp [hunk 1]
```diff
@@ -607,23 +650,26 @@ ConstraintEvaluationResult OrientationConstraint::decide(const moveit::core::Rob
...
+    const Eigen::Vector3d current_rpy = CalcEulerAngles(state.getGlobalLinkTransform(link_model_).linear());
+    xyz = current_rpy - desired_rpy;
   }
-
   // Account for angle wrapping
-  xyz(0) = std::min(fabs(xyz(0)), boost::math::constants::two_pi<double>() - fabs(xyz(0)));
-  xyz(1) = std::min(fabs(xyz(1)), boost::math::constants::two_pi<double>() - fabs(xyz(1)));
-  xyz(2) = std::min(fabs(xyz(2)), boost::math::constants::two_pi<double>() - fabs(xyz(2)));
+  xyz = xyz.unaryExpr([](const double angle) {
+    const double normalized_angle = std::fmod(std::abs(angle), boost::math::constants::two_pi<double>());
```
Comment (JeroenDM) [inline moveit_core/kinematic_constraints/src/kinematic_constraint.cpp:668]: A small nit that was already in the existing code, but would you mind using `namespace bmc = boost::math::constants` or `M_PI` depending on your stance on  #2285. I think the extremely long name for `pi` really makes it harder to reason about the code.

(I know this is a code base wide issue, so feel free to ignore it in this PR.)

File: moveit_core/utils/src/robot_model_test_utils.cpp [hunk 2]
```diff
@@ -110,7 +110,7 @@ RobotModelBuilder::RobotModelBuilder(const std::string& name, const std::string&
 }
 
 void RobotModelBuilder::addChain(const std::string& section, const std::string& type,
-                                 const std::vector<geometry_msgs::Pose>& joint_origins)
+                                 const std::vector<geometry_msgs::Pose>& joint_origins, urdf::Vector3 joint_axis)
```
Comment (JeroenDM) [inline moveit_core/utils/src/robot_model_test_utils.cpp:113]: Nice! I've missed this but it never occurred to me that I could just add it myself :)

File: moveit_core/kinematic_constraints/src/kinematic_constraint.cpp [hunk 3]
```diff
@@ -58,6 +58,63 @@ static double normalizeAngle(double angle)
   return v;
 }
 
+// Normalizes an angles to an absolute angle.
```
Comment (JeroenDM) [inline moveit_core/kinematic_constraints/src/kinematic_constraint.cpp:61]:
```suggestion
// Normalizes an angle to the interval [-pi, +pi] and then take the absolute value
```

I know this explanation is not how the code does it, but I think it better explains what is happening. Or am I wrong in what the code does?

**Comment by JeroenDM** [comment]
+1 For explicitly handling and explaining the singular case, great work! @JafarAbdi 

I was thinking the following case could maybe surprise some users:
- Orientation tolerance of (0, 0, 1) on roll-pitch-yaw angles.
- Robot at the singular case with theta = 0.5.
In this case, the yaw tolerance is not violated by theta, and therefore theta will be interpreted as pure roll.
This results in roll tolerance being violated.

Is this a correct interpretation?

**Comment by JafarAbdi** [comment]
> I used some hardcoded test data [here](https://github.com/JeroenDM/moveit/blob/49df9efcbf81cff25ab6875023d57931869ed773/moveit_core/kinematic_constraints/test/test_orientation_constraints.cpp#L52), generated with a [Python script](https://gist.github.com/JeroenDM/f8ba5f1a60821c4cacf7c137eff6f409). If you think this is a good approach I can add the test for Euler angles to this PR.

Sorry, I missed this comment, I'm okay with adding them here, or if you prefer I could open my PR against your branch to merge them together

**Comment by JafarAbdi** [comment]
> I was thinking the following case could maybe surprise some users:
> 
> * Orientation tolerance of (0, 0, 1) on roll-pitch-yaw angles.
> * Robot at the singular case with theta = 0.5.
>   In this case, the yaw tolerance is not violated by theta, and therefore theta will be interpreted as pure roll.
>   This results in roll tolerance being violated.
> 
> Is this a correct interpretation?
> 
> **Edit**:
> I don't know how often this singular case is relevant in practical problems, but if my interpretation is correct, an alternative could be to check theta against the maximum of the roll and yaw tolerance.

As you can see in the image below given the tolerances this should pass, I added it as a test case

Screenshot from 2020-12-07 15-36-05

Also, I forgot to mention that in the last commits I reverted my previous approach and used the error rotation matrix to check for tolerances since it's easier to handle singularities

File: moveit_core/kinematic_constraints/test/test_orientation_constraints.cpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+  robot_state.setVariablePositions(getJointValues(0.21, boost::math::constants::half_pi<double>(), 0.0));
+  robot_state.update();
+  EXPECT_TRUE(oc.configure(ocm, tf));
+  EXPECT_FALSE(oc.decide(robot_state).satisfied);
+
+  robot_state.setVariablePositions(getJointValues(0.0, boost::math::constants::half_pi<double>(), 0.21));
+  robot_state.update();
+  EXPECT_TRUE(oc.configure(ocm, tf));
+  EXPECT_FALSE(oc.decide(robot_state).satisfied);
+
```
Comment (JeroenDM) [inline moveit_core/kinematic_constraints/test/test_orientation_constraints.cpp:143]:
```suggestion

  // In the non-singular case the absolute_x_axis_tolerance is not violated
  robot_state.setVariablePositions(getJointValues(0.0, boost::math::constants::half_pi<double>() - 0.01, 0.21));
  robot_state.update();
  EXPECT_TRUE(oc.configure(ocm, tf));
  EXPECT_TRUE(oc.decide(robot_state).satisfied);

```

**Comment by JeroenDM** [comment]
> Also, I forgot to mention that in the last commits I reverted my previous approach and used the error rotation matrix to check for tolerances since it's easier to handle singularities

Ah, ok, now I understand it :) I glanced over that change :s
I suggested an extra test case to clarify the behaviour.

> Sorry, I missed this comment, I'm okay with adding them here, or if you prefer I could open my PR against your branch to merge them together

Given you already wrote extensive tests now, I don't' think my other tests would add value here. I will just extend your test file for the new parameterization once this is merged.

**Review by JeroenDM — APPROVED** [review]

File: moveit_core/kinematic_constraints/test/test_orientation_constraints.cpp [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+  ocm.orientation.y = 0.0;
+  ocm.orientation.z = 0.0;
+  ocm.orientation.w = 1.0;
+  ocm.absolute_x_axis_tolerance = 0.2;
+  ocm.absolute_y_axis_tolerance = 2.0;
+  ocm.absolute_z_axis_tolerance = 0.3;
+  ocm.weight = 1.0;
+
+  moveit::core::RobotState robot_state(robot_model_);
+  // Singularity: roll + yaw = theta
```
Comment (JeroenDM) [inline moveit_core/kinematic_constraints/test/test_orientation_constraints.cpp:179]:
```suggestion
// Singularity: roll - yaw = theta
```

Comment (JeroenDM) [reply moveit_core/kinematic_constraints/test/test_orientation_constraints.cpp:179]: Because the tests below use -pi / 2.

**Comment by JeroenDM** [comment]
Also, next time I will properly organize my comments in a review... sorry.

**Comment by AndyZe** [comment]
I just want to make sure you have especially tested several values aroud pitch = pi/2. That's where singularities occur for the Euler angle roll/pitch/yaw convention, I believe. See [here](https://www.astro.rug.nl/software/kapteyn-beta/_downloads/attitude.pdf), page 13.

singularities

File: moveit_core/kinematic_constraints/src/kinematic_constraint.cpp [hunk 6]
```diff
@@ -58,6 +58,58 @@ static double normalizeAngle(double angle)
   return v;
 }
 
+// Normalizes an angle to the interval [-pi, +pi] and then take the absolute value
+// The returned values will be in the following range [0, +pi]
```
Comment (AndyZe) [inline moveit_core/kinematic_constraints/src/kinematic_constraint.cpp:62]: Minor correction:

// Normalizes the absolute value of an angle to [0, +pi]
// Returned values will be in the range [0, +pi]

**Review by AndyZe — APPROVED** [review]
Overall I think a better way to do this would be to convert to a rotation vector because it is unique and does not have singularities. But I'm OK getting this merged in the meantime because it is breaking a lot of CI setups.

**Review by rhaschke — COMMENTED** [review]
Looks reasonable. Will fix #2449 as well.
However, there is a failing (ikfast) test here. Not sure that is related. I retriggered Travis.

**Comment by AndyZe** [comment]
Let's run CI one more time

**Comment by gavanderhoorn** [comment]
Off-topic, but you don't need to close and re-open a PR to rerun Travis: if you don't see a "rerun" or "restart" button on the *Checks* tab of the PR, you can follow the `details` link to the `travis-ci.com` site, login with your Github account and then restart a run there.

**Comment by AndyZe** [comment]
I should have squash-merged  :/

**Comment by rhaschke** [comment]
> I should have squash-merged :/

Yes, indeed. I cleaned up and force-pushed. Please update your devel branches @tylerjw, @AndyZe, @JafarAbdi.

**Comment by rhaschke** [comment]
@AndyZe: You pushed a branch `revert-2434-pr-fix_orientation_constraints` to the main repo. Was that intended to revert the accidental rebase merge? That would have made it even worse (adding even more commits...) If that branch is not needed anymore, please delete it.

**Comment by AndyZe** [comment]
done!

**Pull request merged**